File: src/clients/cliente.js
const net = require('net');
const fs = require('fs');
const readline = require('readline');

// Dirección y puerto del servidor líder
const SERVER_IP = '192.168.18.48'; // Cambiar por la IP del servidor líder en la LAN
const SERVER_PORT = 8877;

// Conectar al servidor
const client = new net.Socket();
client.connect(SERVER_PORT, SERVER_IP, () => {
    console.log('Conectado al servidor líder. Escribe el nombre del archivo:');

    // Nombre del archivo predeterminado (modificar según la ubicación del archivo)
    const fileName = 'C:\\Users\\guisa\\Desktop\\Concurrente-EF\\src\\main\\res\\input.txt';

    // Verificar que el archivo existe
    if (!fs.existsSync(fileName)) {
        console.error('El archivo no existe o no es un archivo válido.');
        client.end();
        return;
    }

    // Crear interfaz para leer desde la consola
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    rl.question('Selecciona el tipo de búsqueda:\n1: Contar ocurrencias de una palabra\n2: Verificar si una palabra existe\n3: Buscar palabras clave repetidas n veces\n', (tipoBusqueda) => {
        rl.question('Escribe la palabra clave para buscar: ', (palabraClave) => {
            let parametroExtra = '';
            if (tipoBusqueda === '3') {
                rl.question('Escribe el número de repeticiones deseadas: ', (param) => {
                    parametroExtra = param;
                    sendSearchRequest(tipoBusqueda, palabraClave, parametroExtra);
                    rl.close();
                });
            } else {
                sendSearchRequest(tipoBusqueda, palabraClave, parametroExtra);
                rl.close();
            }
        });
    });
});

function sendSearchRequest(tipoBusqueda, palabraClave, parametroExtra) {
    // Enviar la configuración de búsqueda al servidor líder
    client.write(tipoBusqueda + '\n');
    client.write(palabraClave + '\n');
    client.write(parametroExtra + '\n');

    // Leer y enviar el contenido del archivo
    const fileStream = fs.createReadStream('C:\\Users\\guisa\\Desktop\\Concurrente-EF\\src\\main\\res\\input.txt');
    fileStream.on('data', (chunk) => {
        client.write(chunk);
    });

    fileStream.on('end', () => {
        client.write('\n'); // Enviar una línea vacía para indicar el final del archivo

        // Leer respuesta del servidor líder
        console.log('Respuesta del servidor líder:');
        client.on('data', (data) => {
            console.log(data.toString());
        });

        client.on('close', () => {
            console.log('Conexión cerrada');
        });
    });
}

client.on('error', (err) => {
    console.error('Error en la comunicación con el servidor líder:', err.message);
});

File: src/clients/cliente.py
import socket

def main():
    server_ip = "192.168.18.48"  # Cambiar por la IP del servidor líder en la LAN
    port = 8877

    try:
        with socket.create_connection((server_ip, port)) as sock:
            input_stream = sock.makefile('r')
            output_stream = sock.makefile('w')

            print("Conectado al servidor líder. Escribe el nombre del archivo:")

            # Leer el nombre del archivo desde la consola (modificar según la ubicación del archivo)
            file_name = "C:\\Users\\guisa\\Desktop\\Concurrente-EF\\src\\main\\res\\input.txt"

            # Verificar que el archivo existe
            try:
                with open(file_name, 'r') as file:
                    pass
            except FileNotFoundError:
                print("El archivo no existe o no es un archivo válido.")
                return

            print("Selecciona el tipo de búsqueda:")
            print("1: Contar ocurrencias de una palabra")
            print("2: Verificar si una palabra existe")
            print("3: Buscar palabras clave repetidas n veces")
            tipo_busqueda = input()

            print("Escribe la palabra clave para buscar:")
            palabra_clave = input()

            parametro_extra = ""
            if tipo_busqueda == "3":
                print("Escribe el número de repeticiones deseadas:")
                parametro_extra = input()

            # Enviar la configuración de búsqueda al servidor líder
            output_stream.write(tipo_busqueda + '\n')
            output_stream.write(palabra_clave + '\n')
            output_stream.write(parametro_extra + '\n')
            output_stream.flush()

            # Leer y enviar el contenido del archivo
            with open(file_name, 'r') as file:
                for line in file:
                    output_stream.write(line)
                output_stream.write('\n')  # Enviar una línea vacía para indicar el final del archivo
                output_stream.flush()

            # Leer respuesta del servidor líder
            print("Respuesta del servidor líder:")
            while True:
                response = input_stream.readline()
                if not response:
                    break
                print(response.strip())

    except Exception as e:
        print("Error en la comunicación con el servidor líder:", e)

if __name__ == "__main__":
    main()
